core/src/main/scala/io/projectglow/vcf/VCFWriterUtils.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019 The Glow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.projectglow.vcf
+
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.types.{ArrayType, StructType}
+
+import io.projectglow.common.GlowLogging
+
+object VCFWriterUtils extends GlowLogging {
+
+  /**
+   * Infer sample IDs from a genomic DataFrame.
+   *
+   * - If there are no genotypes, there are no sample IDs.
+   * - If there are genotypes and sample IDs are all...
+   *     - Missing, the sample IDs are injected from the number of genotypes per row (must be the same per row).
+   *     - Present, the sample IDs are found by unifying sample IDs across all rows.
+   */
+  def inferSampleIdsAndInjectMissing(data: DataFrame): (Seq[String], Boolean) = {
+    val genotypeSchemaOpt = data
+      .schema
+      .find(_.name == "genotypes")
+      .map(_.dataType.asInstanceOf[ArrayType].elementType.asInstanceOf[StructType])
+    if (genotypeSchemaOpt.isEmpty) {
+      logger.warn("No genotypes column, no sample IDs will be inferred.")
```

Comment:
Doesn't seem like this should be a warning. Probably just an INFO or even DEBUG.